pkg/operator2/workload/workload_controller.go

```diff
@@ -0,0 +1,308 @@
+}
+
+func (c *OAuthAPIServerOperator) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *OAuthAPIServerOperator) processNextWorkItem() bool {
+	dsKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(dsKey)
+
+	err := c.sync()
+	if err == nil {
+		c.queue.Forget(dsKey)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", dsKey, err))
+	c.queue.AddRateLimited(dsKey)
+
+	return true
+}
+
+// eventHandler queues the operator to check spec and status
+func (c *OAuthAPIServerOperator) EventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(workQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(workQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(workQueueKey) },
+	}
+}
+
+// canRunOperator checks ManagementState to determine if we can run this operator, probably set by a cluster administrator.
+func (c *OAuthAPIServerOperator) canRunOperator(operatorSpec *operatorv1.OperatorSpec) (bool, error) {
+	switch operatorSpec.ManagementState {
+	case operatorv1.Managed:
+		return true, nil
+	case operatorv1.Unmanaged:
+		return false, nil
+	case operatorv1.Removed:
+		if err := c.kubeClient.CoreV1().Namespaces().Delete(c.targetNamespace, nil); err != nil && !apierrors.IsNotFound(err) {
+			return false, err
+		}
+		return false, nil
+	default:
+		c.eventRecorder.Warningf("ManagementStateUnknown", "Unrecognized operator management state %q", operatorSpec.ManagementState)
+		return false, nil
+	}
+}
+
+// preconditionFulfilled checks if:
+// - kube-apiserver is present and available
+// - TODO: Spec.ObservedConfig.Raw exists and has desired data
+func (c *OAuthAPIServerOperator) preconditionFulfilled(operatorSpec *operatorv1.OperatorSpec) (bool, error) {
+	kubeAPIServerClusterOperator, err := c.openshiftClusterConfigClient.Get("kube-apiserver", metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		kubeAPIServerClusterOperator, err = c.openshiftClusterConfigClient.Get("openshift-kube-apiserver-operator", metav1.GetOptions{})
+	}
+	if apierrors.IsNotFound(err) {
+		message := "clusteroperator/kube-apiserver not found"
+		c.eventRecorder.Warning("PrereqNotReady", message)
+		return false, fmt.Errorf(message)
+	}
+	if err != nil {
+		return false, err
+	}
+	if !clusteroperatorv1helpers.IsStatusConditionTrue(kubeAPIServerClusterOperator.Status.Conditions, "Available") {
+		message := fmt.Sprintf("clusteroperator/%s is not Available", kubeAPIServerClusterOperator.Name)
+		c.eventRecorder.Warning("PrereqNotReady", message)
+		return false, fmt.Errorf(message)
+	}
+
+	// TODO: block until config is obvserved
+	/*if len(operatorSpec.ObservedConfig.Raw) == 0 {
+		klog.Info("Waiting for observed configuration to be available")
+		return false, nil
+	}*/
+
+	return true, nil
+}
+
+func (c *OAuthAPIServerOperator) updateOperatorStatus(workload *appsv1.DaemonSet, errs []error) error {
```

Comment:
I am having a bit of a trouble following everything that's being checked in this function, could you try to split it a bit to smaller separate units, have helpers for each condition type?